File: Python3/Baekjoon02580_Sudoku.py
# Baekjoon02580_Sudoku.py
def isValid( curr, row, col ):
	# Check row and column
	for i in range( 9 ):
		if row != i and board[i][col] == curr:
			return False

		if col != i and board[row][i] == curr:
			return False

	# Check 3x3
	center = ( (row - ( row % 3 ) + 1), (col - ( col % 3 ) + 1) )
	for r in range( center[0] - 1, center[0] + 2 ):
		for c in range( center[1] - 1, center[1] + 2 ):
			if ( r != row or c != col ) and board[r][c] == curr:
				return False

	return True

def FillSudoku( row, col, idx ):
	if idx == size:
		for i in range( 1, 10 ):
			if isValid( i, row, col ):
				board[row][col] = i
				return True
		return False
	else:
		for i in range( 1, 10 ):
			if isValid( i, row, col ):
				board[row][col] = i
				if FillSudoku( *zeros[idx], idx + 1 ):
					return True
				else:
					board[row][col] = 0
		return False

board = [ ]
for _ in range( 9 ):
	board.append( list( map( int, input().split() ) ) )

zeros = [ ]
for row in range( 0, 9 ):
	for col in range( 0, 9 ):
		if board[row][col] == 0:
			zeros.append( ( row, col ) )

size = len( zeros )
FillSudoku( *zeros[0], 1 )
for row in range( 0, 9 ):
	for col in range( 0, 9 ):
		print( board[row][col], end = ' ' )
	print()

File: C++/Baekjoon02580_Sudoku.cpp
// Baekjoon02580_Sudoku.cpp
// https://www.acmicpc.net/problem/2580
#include <iostream>
#include <vector>

using namespace std;

struct Position {
	int row, col;
	Position( int x, int y ) : row(x), col(y) {}
};

int board[9][9];
vector<Position> zeros;

bool isValid( Position curr);
bool FillSudoku( Position curr, int idx );

int main(void) {
	ios_base::sync_with_stdio( false );
	cin.tie( NULL );
	cout.tie( NULL );

	for ( int i = 0; i < 9; i++ )
		for ( int j = 0; j < 9; j++ )
			cin >> board[i][j];

	for ( int i = 0; i < 9; i++ )
		for ( int j = 0; j < 9; j++ )
			if ( board[i][j] == 0 )
				zeros.push_back( Position(i, j) );

	FillSudoku( zeros[0], 1 );
	for ( int i = 0; i < 9; i++ ) {
		for ( int j = 0; j < 9; j++ )
			cout << board[i][j] << ' ';
		cout << '\n';
	}

	return 0;
}

bool isValid( int currVal, Position curr ) {
	// Check row and column;
	for ( int i = 0; i < 9; i++ ) {
		if ( curr.row != i && board[i][curr.col] == currVal )
			return false;

		if ( curr.col != i && board[curr.row][i] == currVal )
			return false;
	}

	// Check 3x3
	Position center = Position( curr.row - ( curr.row % 3 ) + 1, curr.col - ( curr.col % 3 ) + 1 );
	for ( int r = center.row - 1; r < center.row + 2; r++ )
		for ( int c = center.col - 1; c < center.col + 2; c++ )
			if ( ( r != curr.row || c != curr.col ) && board[r][c] == currVal )
				return false;

	return true;
}

bool FillSudoku( Position curr, int idx ) {
	if ( idx == zeros.size() ) {
		for ( int i = 1; i < 10; i++ )
			if ( isValid( i, curr ) ) {
				board[curr.row][curr.col] = i;
				return true;
			}
		return false;
	} else {
		for ( int i = 1; i < 10; i++ )
			if ( isValid( i, curr ) ) {
				board[curr.row][curr.col] = i;
				if ( FillSudoku( zeros[idx], idx + 1 ) )
					return true;
				else
					board[curr.row][curr.col] = 0;
			}
		return false;
	}
}
